### ISSUE TITLE
[processor/transform] Set attribute values using connection context

### ISSUE THREAD

**Issue by ptodev** [open-telemetry/opentelemetry-collector-contrib#33288]
### Component(s)

processor/transform

### Is your feature request related to a problem? Please describe.

The [attribute](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/attributesprocessor/README.md) and [resource](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourceprocessor/README.md) processors have a `from_context` config argument which can be used to set attribute values using connection context:
```
  # Key specifies the attribute to act upon.
- key: <key>
  action: {insert, update, upsert}
  # FromContext specifies the context value to use to populate the attribute value. 
  # If the key is prefixed with `metadata.`, the values are searched
  # in the receiver's transport protocol additional information like gRPC Metadata or HTTP Headers. 
  # If the key is prefixed with `auth.`, the values are searched
  # in the authentication information set by the server authenticator. 
  # Refer to the server authenticator's documentation part of your pipeline for more information about which attributes are available.
  # If the key doesn't exist, no action is performed.
  # If the key has multiple values the values will be joined with `;` separator.
  from_context: <other key>
```

It would be nice to be able to do the same thing with the transform processor. 

### Describe the solution you'd like

The solution should work with all [OTTL contexts](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/transformprocessor#contexts), so probably we need to add a new [OTTL function](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/ottlfuncs/README.md)?

### Additional context

This feature will help reduce the need for `attributes` and `resource` processors.

Also, #18643 suggests that the `transform` processor may replace the `attributes` and `resource` processors. Having this feature would be a prerequisite for that.

**Comment by odubajDT** [comment]
Hey @TylerHelmuth I would like to look at this issue

**Comment by TylerHelmuth** [comment]
@odubajDT @evan-bradley I'd like to talk about this more.  We don't have syntax or a specific transformcontext to represent how to extract something from the connection context.  I'd like to see some proposals on how we'd implement this before making a PR.

**Comment by evan-bradley** [comment]
Could we have that discussion on a draft PR? I would prefer to look at something concrete when considering possible designs. I think the high-level design is fairly straightforward (`ctx` path available as a map), and the implementation is what's under consideration.

**Comment by evan-bradley** [comment]
As an initial proposal, we'll want a `ctx`/`context` path that is required to be keyed by the path parser and calls `(context.Context).Value` with the key, and returns the return value (an `any`-typed value or `nil`).

As for the transform context, the most conventional way to do this would be to add `context.Context` as the first parameter to each `NewTransformContext` function, but this is something we can explore in a concrete implementation. Once the transform context has the context.Context reference, it's just a matter of retrieving it through a path.

If the implementation shows flaws, we can probably pull this off using a converter that relies on an interface implemented by the transform context, but I'd rather save this as a plan B.

**Comment by TylerHelmuth** [comment]
We keep breaking `NewTransformContext` functions as we need access to more data.  Should we add an Option pattern?

**Comment by evan-bradley** [comment]
I think for now it's conceptually simpler for all data to be required. I think we will always have a `context.Context` object that can be used and can avoid the situation where certain context paths don't work in certain components.

In regards to breaking `NewTransformContext`, as non-ergonomic as it is from a function signature perspective, if we want everything to be required I'd rather break this at compilation time. The alternative is we implement a config struct or options pattern that we have to validate at runtime for containing the correct data, which would place the burden on users instead of component authors if something is required but not present.

**Comment by TylerHelmuth** [comment]
Ok lets try with just breaking it.

Are there any security concerns here? It seems likely you could do something like `merge_maps(attributes, ctx, "upsert")`, and expose keys.

**Comment by evan-bradley** [comment]
I dug into it and there's no way to iterate over `context.Context` without reflection. The path parser will need to require that you key the `ctx` path when using it. I think that should protect us from situations like the one you described, since if you ask for a key, it's your responsibility to know what's in it.

The one tricky thing is that any complex key/value types will be basically impossible to do without custom OTTL extensions. I'm not sure how we properly address this situation:

```
ctx := context.WithValue(ctx, StructKey{key: "key"}, CustomStructData{data: contextData})
```

Keying this would be impossible with OTTL, and even if you use e.g. a string key here instead, using the data is impossible unless you can e.g. convert the type to a map. I think this will just have to be a known caveat.

**Comment by TylerHelmuth** [comment]
Requiring an index should be enforceable by the ottl context path function. I agree we should only support string keys.

**Comment by evan-bradley** [comment]
>I agree we should only support string keys.

We don't have to enforce this, `(context.Context).Value` takes an `any` type as a key and we can just pass a user-supplied value directly and return whatever `Value` returns (`nil` or the value). I was just pointing out that non-primitive typed keys won't work out-of-the-box.

Thinking on it, users could add Converters to provide support for complex key types, e.g. `ctx[KeyFunc("key")]` where `KeyFunc` returns whatever type they use to key context values. So I think we should be good to go here.

**Comment by TylerHelmuth** [comment]
In that case, I think we should be extremely strict about NOT adding any `KeyFunc` converters to `pkg/ottlfuncs`.

**Comment by odubajDT** [comment]
One additional question here, if I understand the implementation of attributes processor correctly, there is the possibility to extract context data only from the following context keys:

```
- "metadata."
- "auth."
- "client.address"
```

What in this case means connection context? Do we want to duplicate the exact same functionality or something else as well? Thanks in advance!

**Comment by evan-bradley** [comment]
>What in this case means connection context?

The Collector passes a `context.Context` object down the pipeline alongside a payload, and usually receivers will include information about the network connection inside of this object. 

>Do we want to duplicate the exact same functionality or something else as well?

I think the fact that we have the ability to spread operations over multiple statements means that we don't have to special-case this kind of addressing. For example, we could have the following: `ctx["auth"]["api-token"]` instead of needing to write `auth.api-token` like it would be in the attributes processor.

**Comment by TylerHelmuth** [comment]
@djaglowski and I talked a bit about this at kubecon.  We think we can get away with including the context.Context in the TransformContext and then adding specific paths to each ottl context to support accessing specific things in the context.Context.  Something like `request.http` to get the map off incoming request headers.

**Comment by ronkorland** [comment]
I've opened a PR that adds the `FromContext` function as discussed in this issue: [#41601](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/41601).

I use this function to retrieve data from an internal extension I'm developing, to validate the authentication of incoming requests. I believe it can be useful for others implementing context-based extensions as well.

Would appreciate any feedback or review

**Comment by TylerHelmuth** [comment]
Ultimately I'd still prefer we use an OTTL context that supplies paths to other OTTL contexts as the way to implement this feature. Since the feature is data access, and that is the responsibility of the path/ottlcontext, I'd prefer not to use a function.

**Comment by edmocosta** [comment]
> Ultimately I'd still prefer we use an OTTL context that supplies paths to other OTTL contexts as the way to implement this feature. Since the feature is data access, and that is the responsibility of the path/ottlcontext, I'd prefer not to use a function.

I agree, and reading this issue I think a few concerns we had in the past no longer apply, such as breaking the `NewTransformContext` to add options.

In addition to that, given the OTTL path getter/setters already have access to `context.Context`, I'm wondering if we really need to pass it down on the `NewTransformContext`? Maybe yes so we can have more control over it?

Revisiting the proposed solution:

```
request.http["foo"] # gets value of "foo" from go.opentelemetry.io/collector/client.Metadata
request.grpc["foo"] # gets value of "foo" from google.golang.org/grpc/metadata
request["foo"] # gets value of "foo" from google.golang.org/grpc/metadata, but if not found there, then tries go.opentelemetry.io/collector/client.Metadata
```

I'm not pretty sure if `request` & `http` would always work here, I think I'd go with something more generic like `context` or `client`, with a sub-path for `context.grpc` as it access the data differently.

Finally, we also have https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/39157 open (cc @axw), which seems to be pretty similar, and could be covered by this change as well. 

WDYT?

**Comment by TylerHelmuth** [comment]
Ya that issue and this issue are connected

**Comment by ronkorland** [comment]
I’ve implemented a `FromContext` function that allows retrieving values from the connection context within OTTL expressions. This approach is designed to provide a flexible and reusable way to access connection-scoped attributes, such as HTTP headers or any custom data stored in the context.

If you have a different approach in mind to solve this problem, I’d be happy to take ownership of this task and help implement it accordingly.

**Comment by edmocosta** [comment]
Thank you @ronkorland, OTTL retrieves the underlying data using OTTL [contexts](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl/contexts) (e.g `log`, `metric`, etc), which provides a few functionalities such as path mapping, enums, etc. Functions and converters are normally reserved to manipulate that data, accessing it through the contexts paths. 

We're discussing here how to provide a new OTTL context to access those `context.Context` values, so using them in OTTL statements/conditions would feel more natural, for example: `set(attributes["val"], context.metadata["key"])`,  `set(attributes["val"], true) where context.metadata["foo"] != "bar"`.

**Comment by ronkorland** [comment]
Thanks @edmocosta, I like this direction and I’d be happy to take this task and work on implementing it.

**Comment by jriguera** [comment]
Please consider also the use case described in this issue: #34649 which makes possible to define context from attributes, which allows users to define tenants.
